### PR TITLE
Cache Discord API objects from the gateway

### PR DESCRIPTION
**Pull request by ebroder**
Set up a mongo cache that is updated whenever the Discord gateway client gets
notifications of object updates. In theory, we could use this to replace our own
API fetches of state.

### CODE REVIEW

**Comment by ebroder** [comment]
(I'm planning to switch over the existing code to use this cache in a separate PR, but I think this one is ready to go)

File: imports/lib/schemas/discord_cache.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+export const DiscordCacheCodec = t.type({
+  _id: t.string,
+  createdAt: date,
+  snowflake: t.string,
+  type: t.string,
+  object: t.object,
+});
+
+const DiscordCacheOverrides: Overrides<t.TypeOf<typeof DiscordCacheCodec>> = {
+  createdAt: {
```
Comment (zarvox) [inline imports/lib/schemas/discord_cache.ts:16]: Any reason to omit (or include) an `updatedAt`?

Comment (ebroder) [reply imports/lib/schemas/discord_cache.ts:16]: No strong reason, I can add one.

**Review by zarvox — APPROVED** [review]
LGTM

Looks like the roles cache is a little tricky, because roles are scoped to the guild that owns them, but presumably they contain some sort of back-reference?

**Comment by ebroder** [comment]
> roles are scoped to the guild that owns them, but presumably they contain some sort of back-reference?

They have a `guild` property, if that's what you mean?

**Pull request merged**